glean/src/server_events.rs

```diff
@@ -0,0 +1,309 @@
+    pub app_id: String,
+    // Version of application emitting the event
+    pub app_display_version: String,
+    // Channel to differentiate logs from prod/beta/staging/devel
+    pub app_channel: String,
+}
+
+// Exported type for public method parameters
+// Default impl empty values will be omitted in json from ping struct definition
+#[derive(Default, Serialize, Deserialize)]
+pub struct RequestInfo {
+    pub user_agent: String,
+    pub ip_address: String,
+}
+
+// Struct to construct the glean ping
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ClientInfo {
+    telemetry_sdk_build: String,
+    first_run_date: String,
+    os: String,
+    os_version: String,
+    architecture: String,
+    app_build: String,
+    app_display_version: String,
+    app_channel: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct PingInfo {
+    seq: u32,
+    start_time: String,
+    end_time: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Ping {
+    document_namespace: String,
+    document_type: String,
+    document_version: String,
+    document_id: String,
+    user_agent: Option<String>,
+    ip_address: Option<String>,
+    payload: String,
+}
+
+type Metrics = HashMap<String, HashMap<String, serde_json::Value>>;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct GleanEvent {
+    category: String,
+    name: String,
+    timestamp: i64,
+    extra: HashMap<String, String>,
+}
+
+pub fn new_glean_event(
+    category: &str,
+    name: &str,
+    extra: std::collections::HashMap<String, String>,
+) -> GleanEvent {
+    GleanEvent {
+        category: category.to_string(),
+        name: name.to_string(),
+        timestamp: Utc::now().timestamp_millis(),
+        extra,
+    }
+}
+#[derive(Serialize, Deserialize, Debug)]
+struct PingPayload {
+    client_info: ClientInfo,
+    ping_info: PingInfo,
+    metrics: Metrics,
+    events: Vec<GleanEvent>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct LogEnvelope {
+    // MozLog compliant format. https://wiki.mozilla.org/Firefox/Services/Logging
+    #[serde(rename = "Type")]
+    log_type: String,
+    // MozLog compliant format. https://wiki.mozilla.org/Firefox/Services/Logging
+    #[serde(rename = "Fields")]
+    fields: Ping,
+}
+
+impl GleanEventsLogger {
+    fn create_client_info(&self) -> ClientInfo {
+        // Fields with default values are required in the Glean schema, but not used in server context
+        ClientInfo {
+            telemetry_sdk_build: "glean_parser v15.0.2.dev17+g81fec69a".to_string(),
+            first_run_date: "Unknown".to_string(),
+            os: "Unknown".to_string(),
+            os_version: "Unknown".to_string(),
+            architecture: "Unknown".to_string(),
+            app_build: "Unknown".to_string(),
+            app_display_version: self.app_display_version.clone(),
+            app_channel: self.app_channel.clone(),
+        }
+    }
+
+    fn create_ping_info() -> PingInfo {
```

Comment:
Again, I'd strongly recommend converting this to an `impl Default`.

Reply:
Totally, this one makes perfect sense.